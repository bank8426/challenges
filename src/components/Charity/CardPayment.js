import React,{useState} from 'react'
import { connect } from 'react-redux';
import styled, { keyframes } from 'styled-components';
import * as paymentActions from '../../redux/actions/paymentActions';
import * as messageActions from '../../redux/actions/messageActions';
import { bindActionCreators } from 'redux'
import RadioButton from '../common/RadioButton';
import PropTypes from 'prop-types'

const cardPaymentFadeInKeyframes = keyframes`
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
`

export const StyledCardPayment = styled.div`
  position: absolute;
  top: 0px;
  width: 100%;
  height: 100%;
  background-color: rgba(255,255,255,0.95);
  padding: 15% 20px;
  box-sizing: border-box;
  border-radius: 10px;
  opacity: 0;
  -webkit-animation: ${cardPaymentFadeInKeyframes} 0.2s forwards cubic-bezier(0.2, 0.8, 0.2, 1) 0.05s;
  animation: ${cardPaymentFadeInKeyframes} 0.2s forwards cubic-bezier(0.2, 0.8, 0.2, 1) 0.05s;
`;

const StyledPaymentList = styled.div`
  display: grid;
  grid-template-columns: repeat(5,1fr);
  padding: 0 100px;
  @media (max-width: 1280px) {
    padding: 0 50px;
  }
  @media (max-width: 1024px) {
    padding: 0px;
  }
  @media (max-width: 768px) {
    padding: 0 80px;
  }
`;

const StyledHeader = styled.h2`
  text-align: center;
`;

export const StyledCloseButton = styled.span`
  position: absolute;
  top: 10px;
  right: 10px;
  border-radius: 50%;
  cursor: pointer;
  font-size: 20px;
  padding: 1px 8px;
  color: #1b53f2;
  font-weight: bold;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: 0.4s cubic-bezier(0.2, 0.8, 0.2, 1);
  transition: 0.4s cubic-bezier(0.2, 0.8, 0.2, 1);
  &:hover{
    color: white;
    background-color: #1B95F2;
    border-color: #0062cc;
  }
  &:active {
    color: rgba(255, 255, 255, 0.5);
    background-color: #1B95F2;
    border-color: #0062cc;
  }
`;

const StyledPayButton = styled.button`
  display: block;
  margin: 2em auto;
`;
export const CardPayment = ({index,savePayment,addMessage,removeMessageById,currency,onClose}) => {
  const payments = [10, 20, 50, 100, 500];
  const [selectedAmount, setSelectedAmount] = useState(payments[0])
  const handlePay = () => {
    savePayment(index, selectedAmount, currency).catch(error => {
      const id = Date.now();
      addMessage('Something went wrong please try again later.',id ,true);
      setTimeout(function() {
        removeMessageById(id)
      }, 2000);
    })
  }
  return (
    <StyledCardPayment>
      <StyledCloseButton onClick={() => onClose()}>X</StyledCloseButton>
      <StyledHeader>Select the amount to donate ({currency})</StyledHeader>
      <StyledPaymentList>
        {
          payments.map((amount, j) => (
            <RadioButton key={j}
              name={`payment${index}`} 
              handleChange={() => {setSelectedAmount(amount)}} 
              displayMessage={amount} 
              checked={amount===selectedAmount}/>
          ))
        }
      </StyledPaymentList>
      <StyledPayButton className='primary' onClick={() => handlePay()}>Pay</StyledPayButton>
    </StyledCardPayment>
  )
}

CardPayment.defaultProps ={
  index : 0,
  savePayment : () => {},
  addMessage : () => {},
  removeMessageById : () => {},
  currency : '',
  onClose : () => {},
}

CardPayment.propTypes={
  index:PropTypes.number.isRequired,
  savePayment: PropTypes.func.isRequired,
  addMessage: PropTypes.func.isRequired,
  removeMessageById: PropTypes.func.isRequired,
  currency:PropTypes.string.isRequired,
  onClose: PropTypes.func.isRequired,
}

const mapDispatchToProps = (dispatch) => ({ 
  savePayment: bindActionCreators(paymentActions.savePayment, dispatch),
  addMessage: bindActionCreators(messageActions.addMessage, dispatch),
  removeMessageById: bindActionCreators(messageActions.removeMessageById, dispatch),
})

export default connect(null,mapDispatchToProps)(CardPayment)


