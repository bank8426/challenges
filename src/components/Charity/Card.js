import React from 'react'
import styled,{keyframes} from 'styled-components';
import CardPayment from './CardPayment';
import PropTypes from 'prop-types'

const cardFadeInKeyframe = keyframes`
  0% {
    opacity: 0;
    top: 10px;
  }
  100% {
    opacity: 1;
    top: 0px;
  }
`

const StyledCard = styled.div`
  margin: 0 10px 50px;
  background: #FFFFFF;
  box-shadow: 2px 2px 10px rgba(0, 0, 0, 0.25);
  border-radius: 10px 10px 10px 10px;
  position: relative;
  -webkit-animation: ${cardFadeInKeyframe} 3s forwards cubic-bezier(0.2, 0.8, 0.2, 1) ${props => (props.index/8+.02)+'s'};
  animation: ${cardFadeInKeyframe} 3s forwards cubic-bezier(0.2, 0.8, 0.2, 1) ${props => (props.index/8+.02)+'s'};
  opacity: 0;
  top: 10px;
`;

const StyledImg = styled.div`
  width: 650px;
  overflow: hidden;
  background-image: url(${props => props.src});
  height: 300px;
  border-radius: 10px 10px 0px 0px;
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center;

  @media (max-width: 1366px) {
    width: 600px;
    height: 275px;
  }
  @media (max-width: 1280px) {
    width: 480px;
    height: 230px;
  }
  @media (max-width: 1024px) {
    width: 385px;
    height: 195px;
  }
  @media (max-width: 834px) {
    width: 345px;
    height: 175px;
  }
  @media (max-width: 768px) {
    width: 100%;
    height: 300px;
  }
  @media (max-width: 500px) {
    width: 100%;
    height: 250px;
  }
`;

const StyledH3 = styled.h3`
  margin: 1em 0em 1em 1em;
`;

export const StyledButton = styled.button`
  margin: auto 1em auto 0;
  @media (max-width: 834px) and (min-width: 768px) {
    margin: 0.5em 0.5em 0.5em 0;
    }
`;

const StyledNameLabel = styled.div`
  display: grid;
  grid-template-columns: 1fr max-content;
`;

const Card = ({item,selectedCharity,setSelectCharity,index}) => {
  return (
    <StyledCard index={index}>
      <StyledImg src={process.env.IMAGES_PATH + item.image}/>
      <StyledNameLabel>
        <StyledH3>{item.name}</StyledH3>
        <StyledButton className='primary' onClick={() => setSelectCharity(item.id) }>Donate</StyledButton>
      </StyledNameLabel>
      { selectedCharity===item.id ? 
        <CardPayment 
          index={item.id} 
          currency={item.currency} 
          onClose={() => setSelectCharity(0)}/> 
        : null }
    </StyledCard>
  )
}

Card.defaultProps ={
  item : {
    id : 0,
    name : '',
    image : '',
    currency : '',
  },
  selectedCharity : -1,
  setSelectCharity : () => {},
  index : 0,
}

Card.propTypes={
  item: PropTypes.shape({
    id :PropTypes.number.isRequired,
    name:PropTypes.string.isRequired,
    image:PropTypes.string.isRequired,
    currency:PropTypes.string.isRequired,
  }),
  selectedCharity : PropTypes.number.isRequired,
  setSelectCharity : PropTypes.func.isRequired,
  index : PropTypes.number.isRequired,
}


export default Card
