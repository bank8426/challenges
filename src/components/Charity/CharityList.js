import React,{useState} from 'react'
import styled from 'styled-components';
import Card from './Card';
const StyledCharityList = styled.div`
  display: grid;
  grid-template-columns: repeat(2,1fr);
  justify-items: center;
  @media (max-width: 768px) {
    display: block;
  }
`;

const CharityList = ({charities}) => {
  const [selectedCharity,setSelectCharity] = useState(0)
  return (
    <StyledCharityList>
      {
        charities.map((item, i) => 
          <Card key={i} 
            item={item} 
            selectedCharity={selectedCharity} 
            setSelectCharity={setSelectCharity} />)
      }
    </StyledCharityList>
  )
}


export default CharityList
