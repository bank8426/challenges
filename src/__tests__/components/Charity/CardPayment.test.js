import React from 'react'
import { CardPayment } from '../../../components/Charity/CardPayment'
import renderer from 'react-test-renderer'
import {mount} from 'enzyme'
import 'jest-styled-components'

describe('CardPayment', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(<CardPayment/>)
  })

  it('it should display CharityList', () => {
    const tree = renderer.create(<CardPayment />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should have default savePayment', () => {
    const props = wrapper.props()
    const spy = jest.spyOn(props, 'savePayment');
    props.savePayment()
    expect(spy).toHaveBeenCalledTimes(1)
  })

  it('should have default addMessage', () => {
    const props = wrapper.props()
    const spy = jest.spyOn(props, 'addMessage');
    props.addMessage()
    expect(spy).toHaveBeenCalledTimes(1)
  })

  it('should have default removeMessageById', () => {
    const props = wrapper.props()
    const spy = jest.spyOn(props, 'removeMessageById');
    props.removeMessageById()
    expect(spy).toHaveBeenCalledTimes(1)
  })

  it('should have default onClose', () => {
    const props = wrapper.props()
    const spy = jest.spyOn(props, 'onClose');
    props.onClose()
    expect(spy).toHaveBeenCalledTimes(1)
  })
});