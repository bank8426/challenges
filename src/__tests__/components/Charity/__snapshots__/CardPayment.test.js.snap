// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardPayment it should display CharityList 1`] = `
.c4 {
  display: block;
  position: relative;
  padding-left: 25px;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c4:hover input ~ .checkmark {
  background-color: #ccc;
}

.c4 input:checked ~ .checkmark {
  background-color: #2196F3;
}

.c4 input:checked ~ .checkmark:after {
  display: block;
}

.c4 .checkmark:after {
  margin: 25%;
  width: 50%;
  height: 50%;
  border-radius: 50%;
  background: white;
}

.c5 {
  position: absolute;
  opacity: 0;
  cursor: pointer;
}

.c7 {
  position: absolute;
  top: 0;
  left: 0;
  height: 25px;
  width: 25px;
  background-color: #eee;
  border-radius: 50%;
  -webkit-transition: 0.4s cubic-bezier(0.2,0.8,0.2,1);
  -webkit-transition: 0.4s cubic-bezier(0.2,0.8,0.2,1);
  transition: 0.4s cubic-bezier(0.2,0.8,0.2,1);
}

.c7:after {
  content: "";
  position: absolute;
  display: none;
}

.c6 {
  display: inline;
}

.c0 {
  position: absolute;
  top: 0px;
  width: 100%;
  height: 100%;
  background-color: rgba(255,255,255,0.95);
  padding: 15% 20px;
  box-sizing: border-box;
  border-radius: 10px;
  opacity: 0;
  -webkit-animation: bcCCNc 0.2s forwards cubic-bezier(0.2,0.8,0.2,1) 0.05s;
  -webkit-animation: bcCCNc 0.2s forwards cubic-bezier(0.2,0.8,0.2,1) 0.05s;
  animation: bcCCNc 0.2s forwards cubic-bezier(0.2,0.8,0.2,1) 0.05s;
}

.c3 {
  display: grid;
  grid-template-columns: repeat(5,1fr);
  padding: 0 100px;
}

.c2 {
  text-align: center;
}

.c1 {
  position: absolute;
  top: 10px;
  right: 10px;
  border-radius: 50%;
  cursor: pointer;
  font-size: 20px;
  padding: 1px 8px;
  color: #1b53f2;
  font-weight: bold;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: 0.4s cubic-bezier(0.2,0.8,0.2,1);
  -webkit-transition: 0.4s cubic-bezier(0.2,0.8,0.2,1);
  transition: 0.4s cubic-bezier(0.2,0.8,0.2,1);
}

.c1:hover {
  color: white;
  background-color: #1B95F2;
  border-color: #0062cc;
}

.c1:active {
  color: rgba(255,255,255,0.5);
  background-color: #1B95F2;
  border-color: #0062cc;
}

.c8 {
  display: block;
  margin: 2em auto;
}

@media (max-width:1280px) {
  .c4 {
    font-size: 20px;
  }
}

@media (max-width:1024px) {
  .c4 {
    font-size: 18px;
  }
}

@media (max-width:1280px) {
  .c3 {
    padding: 0 50px;
  }
}

@media (max-width:1024px) {
  .c3 {
    padding: 0px;
  }
}

@media (max-width:768px) {
  .c3 {
    padding: 0 80px;
  }
}

<div
  className="c0"
>
  <span
    className="c1"
    onClick={[Function]}
  >
    X
  </span>
  <h2
    className="c2"
  >
    Select the amount to donate (
    
    )
  </h2>
  <div
    className="c3"
  >
    <label
      className="c4"
    >
      <input
        checked={true}
        className="c5"
        name="payment0"
        onChange={[Function]}
        type="radio"
      />
      <h3
        className="c6"
      >
        10
      </h3>
      <span
        className="checkmark c7"
      />
    </label>
    <label
      className="c4"
    >
      <input
        checked={false}
        className="c5"
        name="payment0"
        onChange={[Function]}
        type="radio"
      />
      <h3
        className="c6"
      >
        20
      </h3>
      <span
        className="checkmark c7"
      />
    </label>
    <label
      className="c4"
    >
      <input
        checked={false}
        className="c5"
        name="payment0"
        onChange={[Function]}
        type="radio"
      />
      <h3
        className="c6"
      >
        50
      </h3>
      <span
        className="checkmark c7"
      />
    </label>
    <label
      className="c4"
    >
      <input
        checked={false}
        className="c5"
        name="payment0"
        onChange={[Function]}
        type="radio"
      />
      <h3
        className="c6"
      >
        100
      </h3>
      <span
        className="checkmark c7"
      />
    </label>
    <label
      className="c4"
    >
      <input
        checked={false}
        className="c5"
        name="payment0"
        onChange={[Function]}
        type="radio"
      />
      <h3
        className="c6"
      >
        500
      </h3>
      <span
        className="checkmark c7"
      />
    </label>
  </div>
  <button
    className="primary c8"
    onClick={[Function]}
  >
    Pay
  </button>
</div>
`;
